pidgin/app.py

```diff
@@ -0,0 +1,113 @@
+import json
+
+import flask
+import requests
+
+
+app = flask.Flask(__name__)
+
+
+# Endpoint to get core metadata as JSON from an object_id.
+@app.route('/json/<path:object_id>') # 'path' allows the use of '/' in the id
+def get_json_metadata(object_id):
+    try:
+        metadata = get_metadata_dict(object_id)
+        return translate_dict_to_json(metadata)
+    except Exception as e:
+        return str(e), 300
+
+# Endpoint to get core metadata as BibTeX from an object_id.
+@app.route('/bibtex/<path:object_id>') # 'path' allows the use of '/' in the id
+def get_bibtex_metadata(object_id):
+    metadata = get_metadata_dict(object_id)
+    return translate_dict_to_bibtex(metadata)
+
+
+# Create a dictionary containing the metadata for a given object_id.
+def get_metadata_dict(object_id):
+    response = request_metadata(object_id) # query to peregrine
+    data = response.get_json()
+    return flatten_dict(data) # translate the response to a dictionary
+
+
+# Translate a dictionary to a JSON string.
+def translate_dict_to_json(d):
+    json_str = json.dumps(d)
+    return json_str
+
+
+# Translate a dictionary to a BibTeX string.
+def translate_dict_to_bibtex(d):
+    bibtex_str = '@misc {' + d['object_id'] + ', '
+    for k, v in d.items(): # add each pair to the BibTeX output
+        bibtex_str += k + ' = "' + str(v) + '", '
+    bibtex_str += '}'
+    return bibtex_str
+
+
+# Flatten a dictionary containing core metadata.
+def flatten_dict(d):
+    flat_d = {}
+    try:
+        data_type = list(d['data'].keys())[0]
+        for k, v in d['data'][data_type][0].items():
+            if k == 'core_metadata_collections':
+                # object_id is unique so the list should only contain one item
+                for k, v in v[0].items():
+                    flat_d[k] = v
+            else:
+                flat_d[k] = v
+    except Exception:
+        error = 'Core metadata not available for this file'
+        if 'errors' in d:
+            error += ': ' + d['errors'][0]
+        raise Exception(error)
+    return flat_d
+
+
+# Get the type of file from the object_id.
+def get_file_type(object_id):
+    query_txt = '{ datanode (object_id: "' + object_id + '") { type } }'
+    response = send_query(query_txt)
+    try:
+        file_type = response.get_json()['data']['datanode'][0]['type']
+    except IndexError:
+        raise Exception('object_id "' + object_id + '" not found')
```

Comment:
Also here, rather than raising the most generic `Exception`, define a custom error class (say, `ObjectNotFound`) and raise that instead. Fence has some good examples in `fence/errors.py` (I think)